### PR TITLE
Diffuse interface support for more IC's and IC user objects

### LINKED ISSUE #13331: More diffuse interfaces for phase field ICs
## Reason
The phase field module requires diffuse interfaces, but currently very few of the ICs allow diffuse interfaces. So phase field simulations start with sharp interfaces, which slows down simulations and can affect the validity of their results.

## Design
I want to add diffuse interface support to the following common IC systems:
- BoundingBoxIC Initial Condition
- PolycrystalCircles User Object
- PolycrystalVoronoi User Object

There is also a minor improvement that can be made to the hyperbolic tangent interface option in SmoothCircleBaseIC.

## Impact
Additional parameter `int_width` for some IC and User Object blocks. Probably some adjustments to existing tests.

### PR DESCRIPTION
**Pull request by itgreenquist**
Diffuse interfaces for the existing IC `BoundingBoxIC` and the Polycrystal IC UserObjects `PolycrystalCircles` and `PolycrystalVoronoi`. Front end changes are limited to the inclusion of a new input parameter for each case: `int_width`. Several tests also had to be modified, especially ones with `int_width` listed as a global parameter.

The hyperbolic profile in `SmoothCircleBaseIC.C` is also modified to more accurately represent the actual interface width. The associated test had to be changed as well.

#closes #13331

### CODE REVIEW

File: framework/include/ics/BoundingBoxIC.h [hunk 1]
```diff
@@ -61,5 +61,7 @@ class BoundingBoxIC : public InitialCondition
 
   /// The Point object constructed from the x2, y2, z2 components for the bottom left BB corner
   const Point _top_right;
+
+  const Real _int_width;
```
Comment (dschwen) [inline framework/include/ics/BoundingBoxIC.h]: `/// interfacial width`

File: framework/src/ics/BoundingBoxIC.C [hunk 2]
```diff
@@ -46,16 +50,30 @@ BoundingBoxIC::BoundingBoxIC(const InputParameters & parameters)
...
 {
 }
 
 Real
 BoundingBoxIC::value(const Point & p)
 {
-  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
-    if (p(i) < _bottom_left(i) || p(i) > _top_right(i))
-      return _outside;
+  if (_int_width <= 0.0)
```
Comment (dschwen) [inline framework/src/ics/BoundingBoxIC.C]: There is no reason to permit negative interface widths, is there? I suggest making the parameter range checked with `"int_width >= 0.0"` and setting this to `== 0.0`

File: framework/src/ics/BoundingBoxIC.C [hunk 3]
```diff
@@ -46,16 +50,30 @@ BoundingBoxIC::BoundingBoxIC(const InputParameters & parameters)
...
+
+    return _inside;
+  }
+  else
+  {
+    Real f_in = 1.0;
+    for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+      if (_bottom_left(i) != _top_right(i))
+        f_in *= 0.5 * (std::tanh(2.0 * (p(i) - _bottom_left(i)) / _int_width) -
+                       std::tanh(libMesh::pi * (p(i) - _top_right(i)) / _int_width));
```
Comment (dschwen) [inline framework/src/ics/BoundingBoxIC.C]: I think this expression needs further explanation. Why the asymmetry between the two sub terms (one has a 2.0 the other one a pi).

File: modules/phase_field/include/userobjects/PolycrystalCircles.h [hunk 4]
```diff
@@ -48,9 +48,13 @@ class PolycrystalCircles : public PolycrystalUserObjectBase
     R
   }; // Names of columns in text file.
   const bool _columnar_3D;
+  Real _int_width;         // Width of diffuse interface
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/PolycrystalCircles.h]: Please convert those to doxygen style comments (`///`) and make it const.

File: modules/phase_field/include/userobjects/PolycrystalCircles.h [hunk 5]
```diff
@@ -48,9 +48,13 @@ class PolycrystalCircles : public PolycrystalUserObjectBase
...
   unsigned int _grain_num; // Number of crystal grains to create
 
   std::vector<Point> _centerpoints; // x,y,z coordinates of circle centers
   std::vector<Real> _radii;         // Radius for each circular grain created
+
+private:
+  Real computeDiffuseInterface(const Point & p, const unsigned int & i) const;
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/PolycrystalCircles.h]: why is this private?

File: modules/phase_field/src/ics/SmoothCircleBaseIC.C [hunk 6]
```diff
@@ -137,8 +137,7 @@ SmoothCircleBaseIC::computeCircleValue(const Point & p, const Point & center, co
     }
 
     case ProfileType::TANH:
-      return (_invalue - _outvalue) * 0.5 *
-                 (std::tanh(libMesh::pi * (radius - dist) / _int_width) + 1.0) +
+      return (_invalue - _outvalue) * 0.5 * (std::tanh(2.0 * (radius - dist) / _int_width) + 1.0) +
```
Comment (dschwen) [inline modules/phase_field/src/ics/SmoothCircleBaseIC.C]: why this change?

Comment (dschwen) [reply modules/phase_field/src/ics/SmoothCircleBaseIC.C]: I see, you are probably defining the interface through the tangent at the mid-point and its intercepts with y=0 and y=1. What I did was matching the derivative at the mid point to the existing cosine interface profile. The previous method makes the interface with more consistent between the COS and TANH profile choices and is what I would prefer.

File: modules/phase_field/src/userobjects/PolycrystalCircles.C [hunk 7]
```diff
@@ -34,13 +35,15 @@ validParams<PolycrystalCircles>()
   params.addParam<std::vector<Real>>("z_positions", "z coordinate for each circle center");
   params.addParam<std::vector<Real>>("radii", "The radius for each circle");
   params.addParam<FileName>("file_name", "File containing circle centers and radii");
+  params.addParam<Real>("int_width", 0, "Width of diffuse interface");
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/PolycrystalCircles.C]: `0.0`

File: modules/phase_field/src/userobjects/PolycrystalVoronoi.C [hunk 8]
```diff
@@ -55,26 +57,37 @@ void
...
+
+  if (_int_width > 0.0)
+    for (MooseIndex(_centerpoints) grain = 0; grain < n_grains; ++grain)
+      if (grain != min_index)
+      {
+        Point next_grain = _centerpoints[grain];
+        Point N = findNormalVector(point, current_grain, next_grain);
+        Point cntr = findCenterPoint(point, current_grain, next_grain);
+        distance = N * (cntr - point);
+        if (distance < 0.6 * _int_width)
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/PolycrystalVoronoi.C]: what is the significance of `0.6` here?

File: modules/phase_field/src/userobjects/PolycrystalVoronoi.C [hunk 9]
```diff
@@ -151,3 +168,60 @@ PolycrystalVoronoi::precomputeGrainStructure()
...
+  Point pa = point + _mesh.minPeriodicVector(_vars[0]->number(), point, p1);
+  Point pb = point + _mesh.minPeriodicVector(_vars[0]->number(), point, p2);
+  return 0.5 * (pa + pb);
+}
+
+Real
+PolycrystalVoronoi::findLinePoint(const Point & point,
+                                  const Point & N,
+                                  const Point & cntr,
+                                  const unsigned int dim) const
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/PolycrystalVoronoi.C]: is `dim` rather a component here?

File: modules/phase_field/src/userobjects/PolycrystalVoronoi.C [hunk 10]
```diff
@@ -151,3 +168,60 @@ PolycrystalVoronoi::precomputeGrainStructure()
...
+{
+  Real Lsum;
+  if (dim == 0)
+    Lsum = N(1) * (point(1) - cntr(1)) + N(2) * (point(2) - cntr(2));
+  else if (dim == 1)
+    Lsum = N(0) * (point(0) - cntr(0)) + N(2) * (point(2) - cntr(2));
+  else
+    Lsum = N(0) * (point(0) - cntr(0)) + N(1) * (point(1) - cntr(1));
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/PolycrystalVoronoi.C]:
```
const Real l_sum =  N((dim+1)%3) * (point((dim+1)%3) - cntr((dim+1)%3)) + N((dim+2)%3) * (point((dim+2)%3) - cntr((dim+2)%3));
```

File: modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i [hunk 11]
```diff
@@ -40,6 +40,7 @@
     type = PolycrystalVoronoi
     rand_seed = 81
     coloring_algorithm = bt
+    int_width = 0
```
Comment (dschwen) [inline modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i]: that's the default, why add it?

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/phase_field/src/ics/SmoothCircleBaseIC.C (hunk 6 above)
Comment (dschwen) [reply modules/phase_field/src/ics/SmoothCircleBaseIC.C]: image

**Comment by dschwen** [comment]
Have you set up a fully symmetric hex grain (metastable configuration) with this IC and verified that you are close to the equilibrium interface shape? It is not immediately obvious to me that your product series of multiple tanh contributions yield an equilibrium interface profile.

File: framework/src/ics/BoundingBoxIC.C (hunk 3 above)
Comment (itgreenquist) [reply framework/src/ics/BoundingBoxIC.C]: Oops. Thanks for catching that.

File: modules/phase_field/src/ics/SmoothCircleBaseIC.C (hunk 6 above)
Comment (itgreenquist) [reply modules/phase_field/src/ics/SmoothCircleBaseIC.C]: [Moelans 2008](https://journals.aps.org/prb/abstract/10.1103/PhysRevB.78.024113) calculated the equilibrium interface profile as this function for symmetric interfaces. That may just be for the free energy she proposed and not necessarily for all free energy functions, but it strikes me as more justifiable than the current form. If you prefer the pi we can keep it that way.

File: modules/phase_field/src/userobjects/PolycrystalVoronoi.C (hunk 8 above)
Comment (itgreenquist) [reply modules/phase_field/src/userobjects/PolycrystalVoronoi.C]: I've been playing with limiting the range to lower the computational cost, but I think the profile works better without this so I am removing it.

File: modules/phase_field/src/userobjects/PolycrystalVoronoi.C (hunk 9 above)
Comment (itgreenquist) [reply modules/phase_field/src/userobjects/PolycrystalVoronoi.C]: Yes

File: modules/phase_field/src/userobjects/PolycrystalVoronoi.C (hunk 10 above)
Comment (itgreenquist) [reply modules/phase_field/src/userobjects/PolycrystalVoronoi.C]: mind_blown

File: modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i (hunk 11 above)
Comment (itgreenquist) [reply modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i]: int_width is in `[GlobalParams]` with a different value

**Comment by itgreenquist** [comment]
> Have you set up a fully symmetric hex grain (metastable configuration) with this IC and verified that you are close to the equilibrium interface shape? It is not immediately obvious to me that your product series of multiple tanh contributions yield an equilibrium interface profile.

I have. The change in the profile is minor and the free energy is nearly equilibrium. Here is a comparison of the free energies using sharp and diffuse interfaces.
hex_energycomp

File: modules/phase_field/src/ics/SmoothCircleBaseIC.C (hunk 6 above)
Comment (dschwen) [reply modules/phase_field/src/ics/SmoothCircleBaseIC.C]: Are you implying that Moelans' _l_ = `int_width` in your formulation? Because that is definitely not the case.

File: modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i (hunk 11 above)
Comment (dschwen) [reply modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i]: why?

File: modules/phase_field/src/ics/SmoothCircleBaseIC.C (hunk 6 above)
Comment (itgreenquist) [reply modules/phase_field/src/ics/SmoothCircleBaseIC.C]: Yeah. She calls it a GB width but it applies to any interface. Why is that not the case?

Comment (dschwen) [reply modules/phase_field/src/ics/SmoothCircleBaseIC.C]: Argh, I messed this up. You are right, the factor should be 2.0 to make it consistent with Moelans' lGb. Let's make that change. I hope it does not break too many inputs...

File: modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i (hunk 11 above)
Comment (itgreenquist) [reply modules/phase_field/test/tests/DeformedGrain/DeformedGrain.i]: I don't know. I looked at the file again and `int_width` is only needed by one other block. I will switch it to specify in that block only.

File: modules/phase_field/src/userobjects/PolycrystalVoronoi.C (hunk 9 above)
Comment (dschwen) [reply modules/phase_field/src/userobjects/PolycrystalVoronoi.C]: I would not call it `dim` then

**Review by dschwen — DISMISSED** [review]

**Comment by dschwen** [comment]
@permcody, why is Bison failing to build with 

```
XXXXXXX.h:64:36: error: redeclaration of 'using ADMaterial<compute_stage>::isBoundaryMaterial'
   using ADMaterial<compute_stage>::isBoundaryMaterial;
```

?? The fix that removes this line was merged into bison in MR !1682 six days ago.

**Review by dschwen — APPROVED** [review]
Looks ok to me. That bison build failure makes no sense.

**Comment by permcody** [comment]
Perhaps Bison hasn’t had a master merge in six 6 days, or at least not at
the time the testing ran. I’m on travel so I may not be able to dig in
right away to look.

**Comment by dschwen** [comment]
Ok, invalidating fixed it, but now it fails on the const Function changes. Should we just go ahead and merge this?

**Comment by permcody** [comment]
Fine with me if the code works for you. We can sort out the integration
stuff ourselves.
